include/camp/resource/cuda.hpp

```diff
@@ -154,31 +156,57 @@ namespace resources
 
       // Memory
       template <typename T>
-      T *allocate(size_t size)
+      T *allocate(size_t size, MemoryAccess ma = MemoryAccess::Device)
       {
         T *ret = nullptr;
         if (size > 0) {
           auto d{device_guard(device)};
-          campCudaErrchk(cudaMallocManaged(&ret, sizeof(T) * size));
+          switch (ma) {
+            case MemoryAccess::Device:
+              campCudaErrchk(cudaMalloc(&ret, sizeof(T) * size));
+              break;
+            case MemoryAccess::Pinned:
+              // TODO: do a test here for whether managed is *actually* shared
+              // so we can use the better performing memory
+              campCudaErrchk(cudaMallocHost(&ret, sizeof(T) * size));
+              break;
```

Comment:
I had originally named it "Shared".  Reason being that while there are three kinds of memory you can allocate: pinned, truly shared, and managed single-access memory like on pascal and earlier, cuda and hip only allow allocating pinned and one of the other two.  Pinned and truly shared provide the same access guarantees, managed does not, so I was tempted to roll pinned and truly shared into one thing, and have managed be the other one.

That would be different from cuda and hip though.

Reply:
That's true, I believe even truly shared memory can't be used with gpu non-aware MPI on sierra though.

Reply:
It should work, but very slowly. If it actually flat doesn't work that would be a potentially serious problem, since it would mean their access guarantees aren't being met.

Reply:
My recollection was that it gives the wrong answer.